### PR TITLE
#12103 - CI for FreeBSD

### LINKED ISSUE #12103: CI for FreeBSD and OpenBSD
### Summary

We currently have CI support for Linux, macOS and Windows. It would be nice to expand this to cover other operating systems.

### Description

We currently have GitLab CI, Travis CI and AppVeyor CI, but none of these providers support FreeBSD or OpenBSD (or any other BSD for that matter).

It would be nice to expand this coverage and possibly provide pre-built binaries for these platforms.

I oppose that we use [builds.sr.ht](https://builds.sr.ht) to provide CI for other platforms as they support both [FreeBSD](https://man.sr.ht/builds.sr.ht/compatibility.md#freebsd) and [OpenBSD](https://man.sr.ht/builds.sr.ht/compatibility.md#openbsd) and are growing rapidly so could include other platforms (and architectures!) in the future. They also have some cool features like being able to SSH in to build VMs to test builds.

### Alternatives

[Cirrus CI](https://cirrus-ci.org) support FreeBSD, but not OpenBSD.

### LINKED ISSUE #12113: Koch bootCI error on FreeBSD
When running `koch runCI` on FreeBSD, the following error is reported:

```
[build@build /usr/home/build/Nim]$ ./koch runCI
runCI:
'/usr/home/build/Nim/koch' boot
sh: Syntax error: Unterminated quoted string
FAILURE
```

When running `echo $SHELL`, the shell is `/usr/local/bin/bash`.

Running `'/usr/home/build/Nim/koch' boot` from the shell works correctly.

### PR DESCRIPTION
**Pull request by euantorano**
Fixes #12113 
Fixes #12103 

Linked to #12105 - this adds FreeBSD CI support where the other adds OpenBSD.

There are a few tests currently failing on FreeBSD. I've opened some issues for some of them and will open more after the current test run completes.

### CODE REVIEW

**Comment by euantorano** [comment]
This PR is complete, the build fails due to other FreeBSD errors that I've opened issues for.

**Comment by euantorano** [comment]
This seems to be hanging at `PASS: tests/dll/client.nim C --gc:boehm --threads:on               ( 0.43747401 secs)` every time. I'm not sure which test should be running after that that's hanging though...

I've tested it locally on FreeBSD 12 RELEASE and it runs through all the tests without a problem, so I'm not sure what's going on.

**Comment by narimiran** [comment]
How can we see what is failing? For you openBSD PR there are errors printed at the end of the log, but I don't see the same here.

**Comment by euantorano** [comment]
@narimiran What's failing here is a timeout (see my previous comment), which I can't replicate on VMs or physical hardware. I'll push an empty commit to this PR later to try trigger another test run and see if it still times out.

**Comment by narimiran** [comment]
> I'll push an empty commit to this PR later

Even better, you can rebase it against the latest devel and solve the existing merge conflict :)

**Comment by euantorano** [comment]
@narimiran Sure, that should be easy enough. Will do so after work.

**Comment by euantorano** [comment]
Looks like it's still stalling at the same point - it reaches `PASS: tests/dll/client.nim C --gc:boehm --threads:on` and then gets stuck...

I have no idea why, or what it's actually meant to be doing at that point.

**Comment by narimiran** [comment]
Could you disable that test so we can see if something else fails too?

The top of `tests/dll/client.nim` should read:
```nim
discard """
  disabled: true
  output: "Done"
  cmd: "nim $target --debuginfo --hints:on --define:useNimRtl $options $file"
"""
```

**Comment by euantorano** [comment]
Yep, will give it a go.

**Comment by euantorano** [comment]
So I disabled `tests/dll/client.nim` and it now logs the following:

```
SKIP: tests/dll/client.nim C --gc:boehm --threads:on
```

It now seems to be stuck again, so whichever test is running after this one seems to be the problem.

File: lib/pure/os.nim [hunk 1]
```diff
@@ -2722,7 +2722,7 @@ when not weirdTarget and (defined(freebsd) or defined(dragonfly)):
           result.setLen(0) # error!
           break
       else:
-        result.setLen(pathLength)
+        result.setLen(pathLength-1)
```
Comment (Araq) [inline lib/pure/os.nim]: Suspicious. Why is this -1 here? At least add a comment.

Comment (euantorano) [reply lib/pure/os.nim]: The `-1` is to trim the trailing null byte. I can certainly add a comment to make that clear.

**Comment by euantorano** [comment]
Oh wow, the CI actually passed! 😮

File: lib/pure/os.nim (hunk 1 above)
Comment (euantorano) [reply lib/pure/os.nim]: Comment added in 4f4fa0c.

**Comment by euantorano** [comment]
Build failures seem unrelated as far as I can see - the current `devel` branch is failing too.

File: lib/pure/os.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/os.nim:2726]: @euantorano you are right but it seems like `result = newString(pathLength)` right above is buggy for several reasons...

the (not well named) newString returns uninitilized data, but in fact should use the data provided

https://www.freebsd.org/cgi/man.cgi?sysctl(3)

> The information is	copied into the	buffer specified by oldp.  The size of
     the buffer	is given by the	location specified by oldlenp before the call,
     and that location gives the amount	of data	copied after a successful call
     and after a call that returns with	the error code ENOMEM.	If the amount
     of	data available is greater than the size	of the buffer supplied,	the
     call supplies as much data	as fits	in the buffer provided and returns
     with the error code ENOMEM.  If the old value is not desired, oldp	and
     oldlenp should be set to NULL.

also it's maybe wasteful to allocate `newString(pathLength)`, instead using a `array` buffer as in most other functions allows only allocating what's needed

Comment (euantorano) [reply lib/pure/os.nim:2726]: @timotheecour Good catch, it seems you are right. There are two options:

- Keep the data from the first call, and grow the buffer
- Delete the old data, and start with a clean buffer (eg: first call is just used to get the required buffer size, then a second call is made to get the actual data)

I think the second option would be easiest to implement. I'll open a new issue for this.

Comment (timotheecour) [reply lib/pure/os.nim:2726]: answering on issue u opened https://github.com/nim-lang/Nim/issues/13806

**Pull request merged**